ServerCore/Pages/Events/SinglePlayerPuzzleStateMap.cshtml.cs

```diff
@@ -0,0 +1,159 @@
+                {
+                    playerIdToPlayerStatsMap[puzzleState.PlayerID] = new PlayerStats { Player = puzzleState.Player };
+                }
+
+                if (puzzleState.SolvedTime != null)
+                {
+                    puzzleIdToPuzzleStatsMap[puzzleState.PuzzleID].SolveCount++;
+                    playerIdToPlayerStatsMap[puzzleState.PlayerID].SolveCount++;
+                    playerIdToPlayerStatsMap[puzzleState.PlayerID].Score += puzzleState.Puzzle.SolveValue;
+                }
+            }
+
+            // Order both lists
+            this.PuzzleStatList = puzzleIdToPuzzleStatsMap
+                .Values
+                .OrderBy(puzzleStat => puzzleStat.SolveCount)
+                .ThenBy(puzzleStat => puzzleStat.Puzzle.Name)
+                .ToList();
+
+            this.PlayerStatList = playerIdToPlayerStatsMap
+                .Values
+                .OrderBy(playerStat => playerStat.Score)
+                .ThenBy(playerStat => playerStat.SolveCount)
+                .ThenBy(playerStat => playerStat.Player.Name)
+                .ToList();
+
+            // Create maps from each puzzle/player id to it's sorted index
+            var index = 0;
+            Dictionary<int, int> puzzleIdToIndexMap = this.PuzzleStatList.ToDictionary(puzzleStat => puzzleStat.Puzzle.ID, puzzleStat => index++);
+
+            index = 0;
+            Dictionary<int, int> playerIdToIndexMap = this.PlayerStatList.ToDictionary(playerStat => playerStat.Player.ID, playerStat => index++);
+
+            // Create final stat matrix
+            this.StateMap = new StateStats[this.PlayerStatList.Count, this.PuzzleStatList.Count];
+            foreach (SinglePlayerPuzzleStatePerPlayer statePerPlayer in puzzleStatesPerPlayer)
+            {
+                int playerIndex = playerIdToIndexMap[statePerPlayer.PlayerID];
+                int puzzleIndex = puzzleIdToIndexMap[statePerPlayer.Puzzle.ID];
+                this.StateMap[playerIndex, puzzleIndex] = new StateStats()
+                {
+                    UnlockedTime = statePerPlayer.UnlockedTime,
+                    SolvedTime = statePerPlayer.SolvedTime,
+                    LockedOut = statePerPlayer.IsLockedOut
+                };
+            }
+
+            return Page();
+        }
+
+        public class PuzzleStats
+        {
+            public Puzzle Puzzle { get; set; }
+            public int SolveCount { get; set; }
+        }
+
+        public class PlayerStats
+        {
+            public PuzzleUser Player { get; set; }
+            public int SolveCount { get; set; }
+            public int Score { get; set; }
+        }
+
+        public class StateStats
+        {
+            public static StateStats Default => new StateStats();
+
+            public DateTime? UnlockedTime { get; set; }
+            public DateTime? SolvedTime { get; set; }
+            public bool LockedOut { get; set; }
+
+            public string Classes
```

Comment:
I get that this is a simplified version of the function of same name in Map.cshtml - but can it be shared, so that if we make updates we wind up fixing them both at once? Maybe StateStats gets shared and the SinglePlayer version just provides default values for some of the properties it doesn't care about?

Reply:
done